### DISCUSSION TITLE
Hot Reload For Server

### DISCUSSION THREAD

**Discussion by tauseefsshah** [clue/framework-x#86] [Ideas]
Just got my hands on the framework and while trying out stuff I felt the need for a hot reloading server because it's a very handy and a convenient feature to have. I feel having to restart the server after every code change is inefficient.

I might have missed it but if there's a way to achieve it, I'll appreciate you sharing.

**Comment by clue** [comment]
@tauseefsshah Thanks for bringing this up, I agree that hot reloading the source code would be a nice feature!

At the moment, the easiest work around is to use PHP's development web server for development purposes and the built-in web server for production use:

```bash
# development (reloads source for every request)
$ php -S localhost:8080 public/index.php

# deployment (keeps everything in memory)
$ php public/index.php
```

X supports running anywhere and we can take advantage of this for development purposes as both modes can be used mostly interchangeably.

Longer term, we're definitely interested in bringing in hot reloading also for the built-in web server. Unfortunately, this is non-trivial as checking the filesystem is inherently blocking and as such rather costly. Periodically spawning a child process (how often?) only for this purpose seems quite excessive; taking advantage of inotify might be option, but either requires an extension or FFI, so I'm not sure how feasible this is. We've also discussed an option to use a filesystem monitor in sytemd to restart the service unit if that's your cup of tea. Many options…

Would love to hear more input on this! :+1:

**Reply by tauseefsshah** [reply]
@clue Thanks. I agree checking the filesystem is costly but it'll be nice to have hot reloading inbuilt for development environment in future. I guess for now I'll go with PHP's web development server or maybe setup `chokidar`  to monitor file changes and restart the server, here's a [blog](https://sebastiandedeyne.com/file-watching-with-chokidar/) for it.

**Reply by staabm** [reply]
I agree that having a hot reload feature would be great.

nevertheless: is there a reason why you don't use the PHP's development web server? maybe improving the experience with the development web server is also an option, in case it is missing something, you didn't mention yet.

**Reply by tauseefsshah** [reply]
@staabm PHP's development web server is awesome but personally for local development I prefer setting up `vhosts` with `.localhost` TLD so that I never have to start or stop the server.

**Reply by staabm** [reply]
ok, so in other words you would prefer a maybe slower feedback loop with hot reloading on every request, over a one time start/stop development server process? I don't judge you for that, just want to make sure I get your point.

**Reply by tauseefsshah** [reply]
> hot reloading on every request

Yes but it hot reloads on every file change not on every request.

**Reply by john-everden** [reply]
I wonder if making a development base docker image and publishing it with fswatch and some configurable options would be sufficient?

**Reply by SimonFrings** [reply]
@john-everden Sounds interesting, more options more better!
We're currently involved in some other features for Framework X but we're always happy about PRs :+1:

**Comment by mahmutbayri** [comment]
I use https://github.com/seregazhuk/php-watcher and It works fine.

**Reply by clue** [reply]
@mahmutbayri Also a great suggestion and great work by @seregazhuk! ❤️

I wonder if it makes sense to implement a similar logic as part of X?

**Reply by mahmutbayri** [reply]
> @mahmutbayri Also a great suggestion and great work by @seregazhuk! ❤️
> 
> I wonder if it makes sense to implement a similar logic as part of X?

I use like that

`X_LISTEN=0.0.0.0:8081 ./vendor/bin/php-watcher --watch src --watch resources public/index.php --ext=php,twig`

it helped me a lot.

**Reply by john-everden** [reply]
Doesn't appear to support PHP 8.1 yet :(

**Reply by mahmutbayri** [reply]
Please install it like this.

`composer global require seregazhuk/php-watcher:dev-master`
or
`composer require seregazhuk/php-watcher:dev-master`

**Comment by clue** [comment]
Small update: I've started working on this and have a working prototype ready:

![carbon(89)](https://user-images.githubusercontent.com/776829/177979185-51ba259f-3ce4-4dff-b8e3-7e9e94c08a21.png)

See https://twitter.com/another_clue/status/1545359689961144320

Current focus is to detect changes for included files which already works just fine. Once changes are detected, the process is restarted. The server socket will temporarily be paused without closing and will be reused in the new process, so there should be zero downtime.

However, we should also properly handle any active connections that may exist when a change is detected. The idea is to soft-close any connections and have a fallback to kill any remaining connections after a second. This is doable for all incoming connections, but similar logic should also trigger for any outgoing connections (think an idle connection to your database). PHP lacks low-level system access, but we may utilize FFI (optional, PHP 7.4+) to close any remaining connections.

I'll try to keep this updated as we make progress :+1:

**Reply by CodeWithKyrian** [reply]
Hi @clue . Any updates on this??

**Reply by clue** [reply]
@CodeWithKyrian Haven't seen much demand for this lately, but would still love to find somebody willing to sponsor this development! Personally, I've been using Framework X with Docker Compose watch mode for a while now with great success, see below: https://github.com/clue/framework-x/discussions/86#discussioncomment-13249209.

**Reply by CodeWithKyrian** [reply]
Unfortunately, I don't think I'm in a position to sponsor just yet (as much as I want to). But I've been looking at ways I can contribute in terms of code though. Thanks for all you do!

**Comment by netcarver** [comment]
What about entr?

From the public/ dir...

```ls | entr -r php index.php```

Should restart X whenever anything in public/ changes.

**Comment by clue** [comment]
Plenty of options here already and to add one more option to the mix: I've been using Docker Compose watch mode in pretty much all my projects for a while now and this has mostly solved my personal need for hot reload directly in Framework X recently:

```yaml
services:
  fx:
    build: .
    ports:
      - "${PORT:-8080}:8080"
    develop:
      watch:
        - path: public/
          action: rebuild
        - path: src/
          action: rebuild
        - path: composer.lock
          action: rebuild
        - path: Dockerfile
          action: rebuild
```

With this example `compose.yaml` and a basic `Dockerfile` (https://framework-x.org/docs/best-practices/deployment/#docker-containers), you can run Framework X in the foreground and have it restart anytime a relevant file changes:

```
docker compose up --build --watch
```

Likewise, this may be adjusted to copy over files without rebuilding or apply watch mode to multiple services and more, see https://docs.docker.com/compose/how-tos/file-watch/ for details.
